### PR TITLE
[core][autoscaler][v2] Reconciler - Ray state reconciler [4/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR adds the ray state reconciliation logic for newly running and dead ray nodes. 

Depends on https://github.com/ray-project/ray/pull/42897

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 1]
```diff
@@ -269,3 +269,53 @@ def get_status_transition_times_ns(
             ts_list.append(status_update.timestamp_ns)
 
         return ts_list
+
+    @classmethod
+    def reachable_from(
```
Comment (scv119) [inline python/ray/autoscaler/v2/instance_manager/common.py:274]: get_reachable_status

File: python/ray/autoscaler/v2/instance_manager/common.py [hunk 2]
```diff
@@ -269,3 +269,53 @@ def get_status_transition_times_ns(
...
+
+    @classmethod
+    def _compute_reachable(cls):
+        """
+        Computes and memorize the from status sets for each status machine with
+        a DFS search.
+        """
+        valid_transitions = cls.get_valid_transitions()
+
+        def dfs(graph, start, visited=None):
```
Comment (scv119) [inline python/ray/autoscaler/v2/instance_manager/common.py:303]: first time see a dfs in a while!

File: python/ray/autoscaler/v2/tests/test_reconciler.py [hunk 3]
```diff
@@ -4,6 +4,125 @@
...
+from ray.autoscaler.v2.instance_manager.reconciler import RayStateReconciler
+from ray.core.generated.autoscaler_pb2 import (
+    ClusterResourceState,
+    NodeState,
+    NodeStatus,
+)
+from ray.core.generated.instance_manager_pb2 import Instance
+
+
+def test_ray_reconciler():
```
Comment (scv119) [inline python/ray/autoscaler/v2/tests/test_reconciler.py:17]: consider separate them into individual tests!

**Review by scv119 — APPROVED** [review]

**Comment by rickyyx** [comment]
- updated with asserts on unknown ray node state.

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 4]
```diff
@@ -434,3 +490,107 @@ def _update_instance_manager(
...
+        Raises:
+            ValueError: If the ray status is unknown.
+        """
+        reconciled_im_status = None
+        if ray_status in [NodeStatus.RUNNING, NodeStatus.IDLE]:
+            reconciled_im_status = IMInstance.RAY_RUNNING
+        elif ray_status == NodeStatus.DEAD:
+            reconciled_im_status = IMInstance.RAY_STOPPED
+        elif ray_status == NodeStatus.DRAINING:
+            reconciled_im_status = IMInstance.RAY_STOPPING
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:583]: Is there a case where we stop without drain?

**Review by jjyao — COMMENTED** [review]
Lg

File: python/ray/autoscaler/v2/instance_manager/reconciler.py (hunk 4 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:583]: yeah, i think for providers (Kuberay) that manage ray itself, we should just terminate the cloud instance. let me add to this PR actually.

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:583]: the preemption case should be handled in this PR: https://github.com/ray-project/ray/pull/43078

**Review by jjyao — APPROVED** [review]

**Pull request merged**